rust/src/fsutil.rs

```diff
@@ -0,0 +1,140 @@
+//! Helpers for working with an ostree filesystem
+
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use camino::{Utf8Path, Utf8PathBuf};
+use ostree_ext::{gio, ostree, prelude::*};
+use std::{collections::HashMap, path::Component};
+
+#[derive(Debug, Clone)]
+pub struct ResolvedOstreePaths {
+    pub path: ostree::RepoFile,
+    pub symlink_target: Option<ostree::RepoFile>,
+}
+
+impl ResolvedOstreePaths {
+    /// If the resolved file is itself a symlink, returns the
+    /// file it points to. Otherwise just returns the resolved file.
+    pub fn real_file(&self) -> &ostree::RepoFile {
+        self.symlink_target.as_ref().unwrap_or(&self.path)
+    }
+}
+
+/// Given a path and an ostree filesystem root, resolves
+/// the path to a real file on the filesystem, including
+/// resolving symlinks in the path's directory tree.
+///
+/// Returns a pair of the resolved path and in the case where
+/// the path points to a symlink, it also includes the resolved
+/// symlink target.
+pub fn resolve_ostree_paths<'a>(
+    path: &Utf8Path,
+    fsroot: &ostree::RepoFile,
+    cache: &'a mut HashMap<Utf8PathBuf, ResolvedOstreePaths>,
+) -> Option<ResolvedOstreePaths> {
+    assert!(path.is_absolute());
+
+    // Recurse until root
+    if path.parent() == None {
+        return Some(ResolvedOstreePaths {
+            path: fsroot.clone(),
+            symlink_target: None,
+        });
+    }
+
+    // Check the cache for this path
+    if let Some(cache_hit) = cache.get(path) {
+        return Some(cache_hit.clone());
+    }
+
+    // Resolve our parent, then resolve ourselves as a direct child
+    if let Some(parent) = resolve_ostree_paths(path.parent().unwrap(), fsroot, cache) {
+        let child_file = parent
+            .real_file()
+            .child(path.file_name().unwrap())
+            .downcast::<ostree::RepoFile>()
+            .unwrap();
+
+        if !child_file.query_exists(gio::Cancellable::NONE) {
+            return None;
+        }
+
+        let child_info = child_file
+            .query_info("*", gio::FileQueryInfoFlags::NONE, gio::Cancellable::NONE)
+            .expect("failed to get fs info");
+
+        // If this path is a symlink, figure out what it points to
+        let remapped_target =
+            if child_info.has_attribute("standard::is-symlink") && child_info.is_symlink() {
+                let link_target = child_info.symlink_target().unwrap();
+
+                // Due to a bug in OSTree's Gio.File implementation, we cannot
+                // just do `parent.resolve_relative_path` here as it doesn't correctly
```

Comment:
Yeah, this was a frustrating find. I couldn't figure out why some paths weren't matching, and then after adding a bunch of debugging statements found it was producing paths like `/usr/lib/something/../../usr/bin/something_else` rather than resolving the `..` paths along the way. The documentation for Gio.File says that `resolve_relative_path` should always produce an absolute path, which I guess it _technically_ is here, but IMO it'd be more correct to resolve the real path. Especially because when you call `query_info` on the resulting file, it doesn't actually resolve the real file info (I was getting `FileType::Unknown`).

I'll try and see if I can cobble together a minified repro and file an issue over on the ostree repo. This logic I added does appear to handle these kinds of paths for us just fine when testing a Silverblue build, but still annoying to have to handle.